### ISSUE TITLE
sentry-debug-meta.properties not included in APKs

### ISSUE THREAD

**Issue by clembou** [getsentry/sentry-java#740]
Hi,

We are on the latest version of gradle (3.4.2), and a recent version of the sentry plugin (1.7.23), on the hosted sentry.io deployment.

Stack traces are not de-obfuscated in the UI. I did some research, and I believe this is because the `sentry-debug-meta.properties` file is not included in the final APKs in the `assets` folder. The only sentry related file I see is `sentry-build.properties` at the APK root.

Mapping files are generated and uploaded correctly to the sentry server (I can see them in the `Debug Symbols` project settings page).

```bash
find . -name sentry-debug-meta.properties
./$appname/build/intermediates/merged_assets/staging/out/sentry-debug-meta.properties
```
shows me that the file is generated correctly during the build, but somehow  it is not part of the final apk.

Related issue:
https://github.com/getsentry/sentry-java/issues/631

there were a couple of interesting bug fixes in the last plugin version, so I tried upgrading our project to v1.7.25, but the file is still missing from the APK.

Any idea on what is going on?

I can send a detailed gradle build log, and a slimmed down version of our build.gradle file via email somewhere if this helps?

Thanks!

**Comment by ninniuz** [comment]
Can you confirm you are building a `staging` buildType? Are you using app bundles?

From your description, the only thing that could be happening is that your APK (or aab) is packaged before the sentry-cli is done with uploading the mapping file and writing the debug meta file into the intermediates assets folder.

**Comment by clembou** [comment]
Sorry about the late reply.

Yes, the above case is a staging build (`./gradlew assembleStaging`). We are not using app bundles. 

after the sentry proguard stage (`persistSentryProguardUuidsForStagingStaging`)  I see a similar step for crashlytics (`crashlyticsUploadDeobsStaging`).

Both stages produce a similar properties file, however only the crashylitics one gets included in the build from a clean workspace.

However I just noticed that if I re run the `assembleStaging` command without `./gradlew clean` beforehand, the file gets included, indicating that your theory is indeed correct (the file gets written after the packaging stage, but on a second build the file already exists, so is packaged up).

Any idea on what could be causing this?

**Comment by ninniuz** [comment]
What version of Gradle are you using? (3.4.2 should refer to AGP and not gradle itself).

Are using executing gradle tasks in parallel? (`org.gradle.parallel=true`)

Since the Sentry plugin states that `persistSentryProguardUuidsForStagingStaging` should be run before the `dex` task, then the only way this could fail is that the plugin is not able to correctly detect such task (some heuristics are used to find it).

Can you check if any of the following task is run and when w.r.t. `persistSentryProguardUuidsForStagingStaging`?
- `transformClassesWithDexForStaging`
- `transformClassesWithDexBuilderForStaging`
- `transformClassesAndDexWithShrinkResForStaging`

**Comment by clembou** [comment]
Gradle version:
```bash
> ./gradlew --version

------------------------------------------------------------
Gradle 5.4.1
------------------------------------------------------------

Build time:   2019-04-26 08:14:42 UTC
Revision:     261d171646b36a6a28d5a19a69676cd098a4c19d

Kotlin:       1.3.21
Groovy:       2.5.4
Ant:          Apache Ant(TM) version 1.9.13 compiled on July 10 2018
JVM:          1.8.0_202 (Oracle Corporation 25.202-b08)
OS:           Mac OS X 10.14.6 x86_64
```
We are using `org.gradle.parallel=true`, however turning this off did not fix the problem.

looking at the tasks, I don't see the ones you mention, but I see:
```
> Task :$appName:preBuild UP-TO-DATE
> Task :$appName:extractProguardFiles
> Task :$appName:preStagingBuild
> Task :$appName:compileStagingAidl NO-SOURCE
> Task :$appName:checkStagingManifest
> Task :$appName:compileStagingRenderscript NO-SOURCE
> Task :$appName:generateStagingBuildConfig
> Task :$appName:mainApkListPersistenceStaging
> Task :$appName:mergeStagingShaders
> Task :$appName:compileStagingShaders
> Task :$appName:generateStagingAssets
> Task :$appName:mergeStagingAssets
> Task :$appName:writeStagingApplicationId
> Task :$appName:processStagingGoogleServices
> Task :$appName:createStagingCompatibleScreenManifests
> Task :$appName:processStagingManifest
> Task :$appName:fabricGenerateResourcesStaging
> Task :$appName:generateStagingResValues
> Task :$appName:generateStagingResources
> Task :$appName:mergeStagingResources
> Task :$appName:processStagingResources
> Task :$appName:kaptGenerateStubsStagingKotlin
> Task :$appName:kaptStagingKotlin
> Task :$appName:compileStagingKotlin
> Task :$appName:prepareLintJar UP-TO-DATE
> Task :$appName:generateStagingSources
> Task :$appName:javaPreCompileStaging
> Task :$appName:compileStagingJavaWithJavac
> Task :$appName:compileStagingSources
> Task :$appName:validateSigningStaging
> Task :$appName:signingConfigWriterStaging
> Task :$appName:addSentryProguardSettingsForStaging
> Task :$appName:processStagingJavaRes NO-SOURCE
> Task :$appName:transformResourcesWithMergeJavaResForStaging
> Task :$appName:transformClassesAndResourcesWithR8ForStaging
> Task :$appName:transformClassesAndResourcesWithR8ForStaging
> Task :$appName:mergeStagingJniLibFolders
> Task :$appName:transformNativeLibsWithMergeJniLibsForStaging
> Task :$appName:packageStaging
> Task :$appName:persistSentryProguardUuidsForStagingStaging
> Task :$appName:crashlyticsStoreDeobsStaging
> Task :$appName:crashlyticsUploadDeobsStaging
> Task :$appName:assembleStaging
```

perhaps this can be R8 related?

**Comment by ninniuz** [comment]
Yes, that's the problem.
In this case the plugin sets `appName:persistSentryProguardUuidsForStagingStaging` as a finalizing step of the proguard task (in this case the R8 task - I forgot R8 includes D8 and dex is not run anymore :( ) but it is possible that `sentry-cli` completes its after assemble is done.

On top of this, I think I am also seeing some problems with the ordering of tasks w/o R8 that I would like to investigate a bit further.

@bruno-garcia need your help here :)

**Comment by clembou** [comment]
I can confirm that we have the same problem on another app, also on the latest Gradle version and Android gradle plugin version, but with R8 disabled (ie using proguard).

**Comment by ninniuz** [comment]
As a workaround, can you try setting a manual dependency in your app's `build.gradle` file between `persistSentryProguardUuidsForStagingStaging` and `assembleStaging`?

That is, something like 
```
afterEvaluate {
	def assembleTask = tasks.findByName("assembleStaging")   
	if (assembleTask != null) {
		assembleTask.dependsOn("persistSentryProguardUuidsForStagingStaging")
	}
}
```

**Comment by clembou** [comment]
I can confirm that the above works as expected for both our apps. Thanks for the workaround! Happy to test a new version of the plugin when you found a proper solution.

**Comment by marandaneto** [comment]
@ninniuz your thoughts seem to be right, `transformClassesAndResourcesWithR8For{Variant}` won't run when `R8` is out there, and this task is a conjunction of `desugaring, shrinking, obfuscating, optimizing, and dexing all in one.` which makes it harder as older AGP versions have `dexing` and `proguard` in separated tasks.

What's about this #804 ? of course it's just a draft, but it seems to be way easier to just get the last task which is `package` and depends on it.

**Comment by marandaneto** [comment]
just a small update: AGP 3.6.x has changed its task name again
from `transformClassesAndResourcesWithR8ForRelease` to `minifyReleaseWithR8`
so, one more fix to do, I'll be looking into it soon, still RC though.

**Comment by marandaneto** [comment]
@clembou it might take a few hours to be available on mavenCentral, but a new version is published, please let us know if you find issues :)
https://github.com/getsentry/sentry-java/releases/tag/v1.7.31_gradle_plugin
